test/integration/udpingress_test.go

```diff
@@ -151,6 +164,217 @@ func TestUDPIngressEssentials(t *testing.T) {
+	assert.NoError(t, err)
+
+	defer func() {
+		t.Logf("cleaning up the service %s", service.Name)
+		assert.NoError(t, env.Cluster().Client().CoreV1().Services(ns.Name).Delete(ctx, service.Name, metav1.DeleteOptions{}))
+	}()
+
+	// Test initial configuration with UDP only
+	t.Log("exposing DNS service via UDPIngress")
+	udp := &kongv1beta1.UDPIngress{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "minudp",
+			Annotations: map[string]string{
+				annotations.IngressClassKey: ingressClass,
+			},
+		},
+		Spec: kongv1beta1.UDPIngressSpec{Rules: []kongv1beta1.UDPIngressRule{
+			{
+				Port: 9999,
+				Backend: v1beta1.IngressBackend{
+					ServiceName: service.Name,
+					ServicePort: int(service.Spec.Ports[0].Port),
+				},
+			},
+		}},
+	}
+	c, err := clientset.NewForConfig(env.Cluster().Config())
+	assert.NoError(t, err)
+	udp, err = c.ConfigurationV1beta1().UDPIngresses(ns.Name).Create(ctx, udp, metav1.CreateOptions{})
+	assert.NoError(t, err)
+
+	defer func() {
+		t.Logf("ensuring UDPIngress %s is cleaned up", udp.Name)
+		if err := c.ConfigurationV1beta1().UDPIngresses(ns.Name).Delete(ctx, udp.Name, metav1.DeleteOptions{}); err != nil {
+			if !errors.IsNotFound(err) {
+				require.NoError(t, err)
+			}
+		}
+	}()
+
+	t.Log("configurating a dns query and clients")
+	query := new(dns.Msg)
+	query.Id = dns.Id()
+	query.Question = make([]dns.Question, 1)
+	query.Question[0] = dns.Question{Name: "kernel.org.", Qtype: dns.TypeA, Qclass: dns.ClassINET}
+	dnsUDPClient := new(dns.Client)
+	dnsTCPClient := dns.Client{Net: "tcp"}
+
+	t.Logf("checking udpingress %s status readiness.", udp.Name)
+	ingCli := c.ConfigurationV1beta1().UDPIngresses(ns.Name)
+	assert.Eventually(t, func() bool {
+		curIng, err := ingCli.Get(ctx, udp.Name, metav1.GetOptions{})
+		if err != nil || curIng == nil {
+			return false
+		}
+		ingresses := curIng.Status.LoadBalancer.Ingress
+		for _, ingress := range ingresses {
+			if len(ingress.Hostname) > 0 || len(ingress.IP) > 0 {
+				t.Logf("udpingress hostname %s or ip %s is ready to redirect traffic.", ingress.Hostname, ingress.IP)
+				return true
+			}
+		}
+		return false
+	}, 120*time.Second, 1*time.Second, true)
+
+	t.Logf("checking DNS to resolve via UDPIngress %s", udp.Name)
+	assert.Eventually(t, func() bool {
+		_, _, err := dnsUDPClient.Exchange(query, fmt.Sprintf("%s:9999", proxyUDPURL.Hostname()))
+		return err == nil
+	}, ingressWait, waitTick)
+
+	// Add a TCPIngress with the same port integer as the TCPIngress, pointing to the same Service
+	t.Log("exposing DNS service via TCPIngress")
+	tcp := &kongv1beta1.TCPIngress{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "mintcp",
+			Annotations: map[string]string{
+				annotations.IngressClassKey: ingressClass,
+			},
+		},
+		Spec: kongv1beta1.TCPIngressSpec{Rules: []kongv1beta1.IngressRule{
+			{
+				Port: 8888,
+				Backend: v1beta1.IngressBackend{
+					ServiceName: service.Name,
+					ServicePort: int(service.Spec.Ports[1].Port),
```

Comment:
Should not this be "Spec.Ports[1].Port" ?

Reply:
No, see up above:

```
192     container := generators.NewContainer("coredns", "coredns/coredns", 53)
193     container.Ports[0].Protocol = corev1.ProtocolUDP
194     container.Ports[0].Name = "dnsudp"
195     container.Ports = append(container.Ports, corev1.ContainerPort{Name: "dnstcp", ContainerPort: 53, Protocol: corev1.ProtocolTCP})
```
The outcome here is that the TCP port is second in the list, so the `NewServiceForDeployment()` call will make it the second port in the Service.

Doesn't really matter since they're identical, but whatever, it's logically consistent.

Reply:
Do you mind clarify "they're identical" in the code ? Shouldn't the "port" number be exact the same ?

Reply:
Yes. `service` will have two ports in its spec, the first with `port: 53` and `proto: UDP`, the second with `port: 53` and `proto: TCP`. The TCPIngress only cares about the port integer, since the proto is inherent to the TCPIngress, and thus you could fill this in with the port from `service.Spec.Ports[0]` because it's also 53.